core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -856,16 +869,31 @@ class ReplicaManager(val config: KafkaConfig,
       logReadResultMap.put(topicPartition, logReadResult)
     }
 
+    val allPartitionsNeedHwUpdate: Boolean = isFromFollower &&
+      logReadResults.forall {
```

Comment:
Wouldn't we want to respond if _any_ of the high watermarks need updating?

Also, we can probably turn this into a `var` and check it in the pass above where we are accumulating `bytesReadable`.